File: swift5/Sources/swift5/generating/builder/MainBuilder.swift

fileprivate class Headphone : CustomStringConvertible {
    private let type: String
    private let loudspeakerType: String
    private let bodyType: String
    private let wireType: String

    init(type: String, loudspeakerType: String, bodyType: String, wireType: String) {
        self.type = type
        self.loudspeakerType = loudspeakerType
        self.bodyType = bodyType
        self.wireType = wireType
    }

    var description: String {
        return ("Headphone [type=" + type + ", loudspeakerType=" + loudspeakerType + ", bodyType=" + bodyType
                + ", wireType=" + wireType + "]")
    }
}

fileprivate protocol HeadphoneParts {
    func setType() -> HeadphoneParts
    func setLoudspeaker(_ loudspeakerType: String) -> HeadphoneParts
    func setBody(_ bodyType: String) -> HeadphoneParts
    func setWire(_ wireType: String) -> HeadphoneParts
    func build() -> Headphone
}

fileprivate class HiFiHeadphoneParts : HeadphoneParts {
    private var type: String?
    private var loudspeakerType: String?
    private var bodyType: String?
    private var wireType: String?

    func setType() -> HeadphoneParts {
        self.type = "Hi-Fi"
        return self
    }

    func setLoudspeaker(_ loudspeakerType: String) -> HeadphoneParts {
        self.loudspeakerType = loudspeakerType
        return self
    }

    func setBody(_ bodyType: String) -> HeadphoneParts {
        self.bodyType = bodyType
        return self
    }

    func setWire(_ wireType: String) -> HeadphoneParts {
        self.wireType = wireType
        return self
    }

    func build() -> Headphone {
        return Headphone(type: self.type!, loudspeakerType: self.loudspeakerType!, bodyType: self.bodyType!, wireType: self.wireType!)
    }
}

fileprivate class HiEndHeadphoneParts : HeadphoneParts {
    private var type: String?
    private var loudspeakerType: String?
    private var bodyType: String?
    private var wireType: String?

    func setType() -> HeadphoneParts {
        self.type = "Hi-End"
        return self
    }

    func setLoudspeaker(_ loudspeakerType: String) -> HeadphoneParts {
        self.loudspeakerType = loudspeakerType
        return self
    }

    func setBody(_ bodyType: String) -> HeadphoneParts {
        self.bodyType = bodyType
        return self
    }

    func setWire(_ wireType: String) -> HeadphoneParts {
        self.wireType = wireType
        return self
    }

    func build() -> Headphone {
        return Headphone(type: self.type!, loudspeakerType: self.loudspeakerType!, bodyType: self.bodyType!, wireType: self.wireType!)
    }
}

fileprivate class Engineer {
	private var headphoneParts: HeadphoneParts?;

	public init(headphoneParts: HeadphoneParts) {
		self.headphoneParts = headphoneParts;
	}
	
	public func buildHeadphone() -> Headphone? {
		if (self.headphoneParts is HiFiHeadphoneParts) {
			return self.headphoneParts!
					.setType()
					.setLoudspeaker("First class")
					.setBody("First class")
					.setWire("First class")
					.build();
		} else if (self.headphoneParts is HiEndHeadphoneParts) {
			return self.headphoneParts!
					.setType()
					.setLoudspeaker("Premium class")
					.setBody("Premium class")
					.setWire("Premium class")
					.build();
		} else {
			return nil;
		}
	}
}

public class MainBuilder {
	public static func main() {
		let engineer1: Engineer = Engineer(headphoneParts: HiFiHeadphoneParts());
		let hifi = engineer1.buildHeadphone()!;
		print(hifi);
		let engineer2: Engineer = Engineer(headphoneParts: HiEndHeadphoneParts());
		let hiend = engineer2.buildHeadphone()!;
		print(hiend);
	}
}

File: python/generating/builder/MainBuilder.py
from dataclasses import dataclass
from abc import ABC, abstractmethod

@dataclass
class Headphone:
    type: str
    loudspeakerType: str
    bodyType: str
    wireType: str

class HeadphoneParts(ABC):
    def setType(self):
        pass
    def setLoudspeaker(self, loudspeakerType: str):
        pass
    def setBody(self, bodyType: str):
        pass
    def setWire(self, wireType: str):
        pass
    def build(self):
        pass

class HiFiHeadphoneParts(HeadphoneParts):
    __type: str
    __loudspeakerType: str
    __bodyType: str
    __wireType: str

    def setType(self):
        self.__type = "Hi-Fi"
        return self
    def setLoudspeaker(self, loudspeakerType):
        self.__loudspeakerType = loudspeakerType
        return self
    def setBody(self, bodyType):
        self.__bodyType = bodyType
        return self
    def setWire(self, wireType):
        self.__wireType = wireType
        return self
    def build(self):
        return Headphone(type = self.__type, loudspeakerType = self.__loudspeakerType, bodyType = self.__bodyType, wireType = self.__wireType)

class HiEndHeadphoneParts(HeadphoneParts):
    __type: str
    __loudspeakerType: str
    __bodyType: str
    __wireType: str

    def setType(self):
        self.__type = "Hi-End"
        return self
    def setLoudspeaker(self, loudspeakerType):
        self.__loudspeakerType = loudspeakerType
        return self
    def setBody(self, bodyType):
        self.__bodyType = bodyType
        return self
    def setWire(self, wireType):
        self.__wireType = wireType
        return self
    def build(self):
        return Headphone(type = self.__type, loudspeakerType = self.__loudspeakerType, bodyType = self.__bodyType, wireType = self.__wireType)

class Engineer:
    __headphoneParts: HeadphoneParts

    def __init__(self, headphoneParts):
        self.__headphoneParts = headphoneParts
    def buildHeadphone(self):
        if (isinstance(self.__headphoneParts, HiFiHeadphoneParts)):
            return self.__headphoneParts\
                .setType()\
                .setLoudspeaker("First class")\
                .setBody("First class")\
                .setWire("First class")\
                .build()
        elif (isinstance(self.__headphoneParts, HiEndHeadphoneParts)):
            return self.__headphoneParts\
                .setType() \
                .setLoudspeaker("Premium class")\
                .setBody("Premium class")\
                .setWire("Premium class")\
                .build()
        else:
            return None

def  main():
    engineer1 = Engineer(HiFiHeadphoneParts())
    hifi = engineer1.buildHeadphone()
    print(hifi)
    engineer2 = Engineer(HiEndHeadphoneParts())
    hiend = engineer2.buildHeadphone()
    print(hiend)

main()

